insights/parsers/ssl_certificate.py

```diff
@@ -0,0 +1,188 @@
+==========================================================================================================
+"""
+
+from insights import parser, CommandParser
+from datetime import datetime
+from insights.parsers import ParseException
+from insights.specs import Specs
+from insights.parsers.certificates_enddate import CertificatesEnddate
+
+
+def parse_openssl_output(content):
+    """
+    It parses the output of "openssl -in <single_certificate_file> -xxx".
+    Currently it only supports the attributes which the output is in
+    key=value pairs. It saves the cert info into a dict. The value of notBefore and
+    notAfter are saved to an instance of ExpirationDate, which contains the date
+    in string and datetime format.
+
+    Raises:
+        ParseException: when the output isn't in key=value format or
+                        the notAfter or notBefore isn't expected format.
+    """
+    date_format = '%b %d %H:%M:%S %Y'
+    data = {}
+    for line in content:
+        if '=' not in line:
+            raise ParseException('The line %s is not in key=value format' % line)
+        key, value = [item.strip() for item in line.split('=', 1)]
+        if key in ['notBefore', 'notAfter']:
+            value_without_tz = value.rsplit(" ", 1)[0]
+            try:
+                date_time = datetime.strptime(value_without_tz, date_format)
+            except Exception:
+                raise ParseException('The %s is not in %s format.' % (key, date_format))
+            value = CertificatesEnddate.ExpirationDate(value_without_tz, date_time)
+        data[key] = value
+    return data
+
+
+class CertificateInfo(CommandParser, dict):
+    """
+    Base class to parse the output of "openssl -in <single_certificate_file> -xxx".
+    Currently it only supports the attributes which the output is in
+    key=value pairs.
+
+    Sample Output::
+
+        issuer= /C=US/ST=North Carolina/L=Raleigh/O=Katello/OU=SomeOrgUnit/CN=a.b.c.com
+        notBefore=Dec  7 07:02:33 2020 GMT
+        notAfter=Jan 18 07:02:33 2038 GMT
+        subject= /C=US/ST=North Carolina/L=Raleigh/O=Katello/OU=SomeOrgUnit/CN=a.b.c.com
+
+    Examples:
+        >>> type(cert)
+        <class 'insights.parsers.ssl_certificate.CertificateInfo'>
+        >>> 'issuer' in cert
+        True
+        >>> cert['issuer']
+        '/C=US/ST=North Carolina/L=Raleigh/O=Katello/OU=SomeOrgUnit/CN=a.b.c.com'
+        >>> cert['notBefore'].str
+        'Dec  7 07:02:33 2020'
+    """
+
+    def __init__(self, context):
+        super(CertificateInfo, self).__init__(
+            context,
+            extra_bad_lines=['error opening certificate', 'unable to load certificate'])
+
+    def parse_content(self, content):
+        """
+        This uses the :py:func:`insights.parsers.ssl_certificate.parse_openssl_output` function.
+        See its documentation for parsing details.
+        """
+
+        self.update(parse_openssl_output(content))
```

Comment:
If it's possible that the result of `parse_openssl_output` an empty dict, better skip it by rating a SkipException.

Reply:
@xiangce, updated, thanks.